File: src/lib/torrent/magnet.ts
const HEX40 = /^[a-fA-F0-9]{40}$/;
const BASE32_32 = /^[a-zA-Z2-7]{32}$/;

export type ParsedMagnet = {
  infoHash: string;
  name: string | null;
  trackers: string[];
};

export function isMagnetInput(value: string): boolean {
  const v = value.trim();
  if (v.toLowerCase().startsWith("magnet:")) return true;
  return HEX40.test(v) || BASE32_32.test(v);
}



export function isDirectVideoUrl(value: string): boolean {
  const v = value.trim();
  if (!/^https?:\/\//i.test(v)) return false;
  try {
    new URL(v);
    return true;
  } catch {
    return false;
  }
}

const WEB_READY_EXT = /\.(mp4|m4v|webm)(\?|#|$)/i;
const NOT_WEB_READY_EXT = /\.(mkv|avi|ts|m3u8|mpd|flv|wmv|mpg|mpeg|m2ts)(\?|#|$)/i;

export function directUrlNotWebReady(value: string): boolean {
  try {
    const path = new URL(value.trim()).pathname;
    if (NOT_WEB_READY_EXT.test(path)) return true;
    if (WEB_READY_EXT.test(path)) return false;
    return false;
  } catch {
    return false;
  }
}

export function parseMagnet(value: string): ParsedMagnet | null {
  const v = value.trim();
  if (HEX40.test(v)) return { infoHash: v.toLowerCase(), name: null, trackers: [] };
  if (BASE32_32.test(v)) {
    const hex = base32ToHex(v);
    return hex ? { infoHash: hex, name: null, trackers: [] } : null;
  }
  if (!v.toLowerCase().startsWith("magnet:")) return null;

  let params: URLSearchParams;
  try {
    params = new URL(v).searchParams;
  } catch {
    params = new URLSearchParams(v.slice(v.indexOf("?") + 1));
  }

  let infoHash: string | null = null;
  for (const xt of params.getAll("xt")) {
    const m = xt.match(/urn:btih:([a-zA-Z0-9]+)/i);
    if (!m) continue;
    const raw = m[1];
    if (HEX40.test(raw)) {
      infoHash = raw.toLowerCase();
      break;
    }
    if (BASE32_32.test(raw)) {
      const hex = base32ToHex(raw);
      if (hex) {
        infoHash = hex;
        break;
      }
    }
  }
  if (!infoHash) return null;

  const name = params.get("dn");
  return {
    infoHash,
    name: name ? decodeURIComponent(name.replace(/\+/g, " ")) : null,
    trackers: params.getAll("tr").filter(Boolean),
  };
}

export function infoHashFromUrl(url: string): { infoHash: string; fileIdx?: number } | null {
  const m = url.match(/(?:^|[/=])([a-fA-F0-9]{40})(?:\/(\d+))?(?=[/?#]|$)/);
  if (!m) return null;
  const fileIdx = m[2] != null ? Number(m[2]) : undefined;
  return {
    infoHash: m[1].toLowerCase(),
    fileIdx: fileIdx != null && Number.isFinite(fileIdx) ? fileIdx : undefined,
  };
}

export function infoHashFromSources(sources?: string[]): string | null {
  if (!sources) return null;
  for (const s of sources) {
    const m = s.match(/^dht:([a-fA-F0-9]{40})$/i);
    if (m) return m[1].toLowerCase();
  }
  return null;
}

function base32ToHex(input: string): string | null {
  const alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";
  const clean = input.toUpperCase().replace(/=+$/, "");
  let bits = "";
  for (const ch of clean) {
    const idx = alphabet.indexOf(ch);
    if (idx === -1) return null;
    bits += idx.toString(2).padStart(5, "0");
  }
  let hex = "";
  for (let i = 0; i + 8 <= bits.length; i += 8) {
    hex += parseInt(bits.slice(i, i + 8), 2).toString(16).padStart(2, "0");
  }
  return hex.length === 40 ? hex : null;
}

File: clients/lib/domain/streams/magnet.dart
/// Magnet / info-hash utilities, ported 1:1 from `src/lib/torrent/magnet.ts`.
/// Used by the stream fetch layer (to recover an info-hash for uncached
/// torrents that omit it) and by stream resolution.
library;

final RegExp _hex40 = RegExp(r'^[a-fA-F0-9]{40}$');
final RegExp _base32_32 = RegExp(r'^[a-zA-Z2-7]{32}$');

/// A parsed magnet: the 40-hex info-hash plus optional display name/trackers.
class ParsedMagnet {
  const ParsedMagnet({
    required this.infoHash,
    required this.name,
    required this.trackers,
  });

  final String infoHash;
  final String? name;
  final List<String> trackers;
}

/// Whether [value] is a magnet URI or a bare 40-hex / 32-base32 info-hash.
bool isMagnetInput(String value) {
  final v = value.trim();
  if (v.toLowerCase().startsWith('magnet:')) return true;
  return _hex40.hasMatch(v) || _base32_32.hasMatch(v);
}

final RegExp _videoUrlExt = RegExp(
  r'\.(mp4|m4v|mkv|webm|mov|avi|ts|m3u8|mpd|flv|wmv|mpg|mpeg|m2ts)(\?|#|$)',
  caseSensitive: false,
);

/// Whether [value] is an http(s) URL whose path ends in a known video
/// container/extension.
bool isDirectVideoUrl(String value) {
  final v = value.trim();
  if (!RegExp(r'^https?://', caseSensitive: false).hasMatch(v)) return false;
  final uri = Uri.tryParse(v);
  if (uri == null) return false;
  return _videoUrlExt.hasMatch(uri.path);
}

final RegExp _webReadyExt = RegExp(
  r'\.(mp4|m4v|webm)(\?|#|$)',
  caseSensitive: false,
);

/// Whether a direct URL is *not* a browser-web-ready container (mp4/m4v/webm).
/// Unparseable URLs are treated as not-web-ready.
bool directUrlNotWebReady(String value) {
  final uri = Uri.tryParse(value.trim());
  if (uri == null) return true;
  return !_webReadyExt.hasMatch(uri.path);
}

/// Parses a magnet URI (or bare hash) into its info-hash, name, and trackers.
/// Returns null if no valid info-hash is present.
ParsedMagnet? parseMagnet(String value) {
  final v = value.trim();
  if (_hex40.hasMatch(v)) {
    return ParsedMagnet(
      infoHash: v.toLowerCase(),
      name: null,
      trackers: const [],
    );
  }
  if (_base32_32.hasMatch(v)) {
    final hex = _base32ToHex(v);
    return hex == null
        ? null
        : ParsedMagnet(infoHash: hex, name: null, trackers: const []);
  }
  if (!v.toLowerCase().startsWith('magnet:')) return null;

  Map<String, List<String>> params;
  final uri = Uri.tryParse(v);
  if (uri != null) {
    params = uri.queryParametersAll;
  } else {
    final qIdx = v.indexOf('?');
    params = Uri(
      query: qIdx >= 0 ? v.substring(qIdx + 1) : '',
    ).queryParametersAll;
  }

  String? infoHash;
  for (final xt in params['xt'] ?? const <String>[]) {
    final m = RegExp(
      r'urn:btih:([a-zA-Z0-9]+)',
      caseSensitive: false,
    ).firstMatch(xt);
    if (m == null) continue;
    final raw = m.group(1)!;
    if (_hex40.hasMatch(raw)) {
      infoHash = raw.toLowerCase();
      break;
    }
    if (_base32_32.hasMatch(raw)) {
      final hex = _base32ToHex(raw);
      if (hex != null) {
        infoHash = hex;
        break;
      }
    }
  }
  if (infoHash == null) return null;

  final dn = params['dn'];
  final name = (dn != null && dn.isNotEmpty) ? dn.first : null;
  return ParsedMagnet(
    infoHash: infoHash,
    name: name != null ? Uri.decodeComponent(name.replaceAll('+', ' ')) : null,
    trackers: (params['tr'] ?? const <String>[])
        .where((t) => t.isNotEmpty)
        .toList(),
  );
}

/// Extracts a 40-hex info-hash (and optional file index) embedded in a URL
/// path, e.g. `.../<hash>/<fileIdx>`.
({String infoHash, int? fileIdx})? infoHashFromUrl(String url) {
  final m = RegExp(
    r'(?:^|[/=])([a-fA-F0-9]{40})(?:/(\d+))?(?=[/?#]|$)',
  ).firstMatch(url);
  if (m == null) return null;
  final idxStr = m.group(2);
  final fileIdx = idxStr != null ? int.tryParse(idxStr) : null;
  return (infoHash: m.group(1)!.toLowerCase(), fileIdx: fileIdx);
}

/// Extracts an info-hash from a Stremio `sources` list (`dht:<hash>` entries).
String? infoHashFromSources(List<String>? sources) {
  if (sources == null) return null;
  for (final s in sources) {
    final m = RegExp(
      r'^dht:([a-fA-F0-9]{40})$',
      caseSensitive: false,
    ).firstMatch(s);
    if (m != null) return m.group(1)!.toLowerCase();
  }
  return null;
}

String? _base32ToHex(String input) {
  const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';
  final clean = input.toUpperCase().replaceAll(RegExp(r'=+$'), '');
  final bits = StringBuffer();
  for (final ch in clean.split('')) {
    final idx = alphabet.indexOf(ch);
    if (idx == -1) return null;
    bits.write(idx.toRadixString(2).padLeft(5, '0'));
  }
  final bitStr = bits.toString();
  final hex = StringBuffer();
  for (var i = 0; i + 8 <= bitStr.length; i += 8) {
    hex.write(
      int.parse(
        bitStr.substring(i, i + 8),
        radix: 2,
      ).toRadixString(16).padLeft(2, '0'),
    );
  }
  final out = hex.toString();
  return out.length == 40 ? out : null;
}
